### PR TITLE
NAS-111965 / 22.12 / APPS: Add bulk upgrade action

### PR DESCRIPTION
**Pull request by denysbutenko**
Changes:

- Add `Upgrade` to `Bulk Actions` menu
- Disable `Bulk Actions` when no selectable items
- Do not close the menu on `Select All` / `Unselect All`

Testing requires a few apps that have updates.

Or use a mock response for `chart.release.query` to see apps with updates and failures while upgrading.
```
midclt call test.set_mock chart.release.query '{"return_value": [{"name":"test-pihole","version":1,"namespace":"ix-test-pihole","id":"test-pihole","catalog":"OFFICIAL","catalog_train":"charts","path":"/mnt/tank/ix-applications/releases/test-pihole","dataset":"tank/ix-applications/releases/test-pihole","status":"ACTIVE","history":{},"update_available":true,"human_version":"2022.10_1.0.7","human_latest_version":"2022.10_1.0.8","pod_status":{"desired":1,"available":1},"used_ports":[],"chart_metadata":{"icon":"https://github.com/pi-hole/graphics/raw/master/Vortex/Vortex_Vertical_wordmark_lightmode.png?raw=true"},"container_images_update_available":false,"item_update_available":true},{"name":"test-nextcloud","version":1,"namespace":"ix-test-nextcloud","id":"test-nextcloud","catalog":"OFFICIAL","catalog_train":"charts","path":"/mnt/tank/ix-applications/releases/test-nextcloud","dataset":"tank/ix-applications/releases/test-nextcloud","status":"ACTIVE","history":{},"update_available":false,"human_version":"25_1.6.15","human_latest_version":"25_1.6.16","pod_status":{"desired":2,"available":2},"used_ports":[],"chart_metadata":{"icon":"https://cdn.rawgit.com/docker-library/docs/defa5ffc7123177acd60ddef6e16bddf694cc35f/nextcloud/logo.svg"},"container_images_update_available":true,"item_update_available":true},{"name":"true-nextcloud","version":1,"namespace":"ix-true-nextcloud","id":"true-nextcloud","catalog":"TRUECHARTS","catalog_train":"stable","path":"/mnt/tank/ix-applications/releases/true-nextcloud","dataset":"tank/ix-applications/releases/true-nextcloud","status":"ACTIVE","history":{},"update_available":true,"human_version":"24.0.6_15.3.34","human_latest_version":"24.0.6_15.3.36","pod_status":{"desired":1,"available":1},"used_ports":[],"chart_metadata":{"icon":"https://cdn.rawgit.com/docker-library/docs/defa5ffc7123177acd60ddef6e16bddf694cc35f/nextcloud/logo.svg"},"container_images_update_available":false,"item_update_available":true},{"name":"true-pihole","version":1,"namespace":"ix-true-pihole","id":"true-pihole","catalog":"OFFICIAL","catalog_train":"charts","path":"/mnt/tank/ix-applications/releases/true-pihole","dataset":"tank/ix-applications/releases/true-pihole","status":"ACTIVE","history":{},"update_available":false,"human_version":"2022.10_1.0.7","human_latest_version":"2022.10_1.0.8","pod_status":{"desired":4,"available":4},"used_ports":[],"chart_metadata":{"icon":"https://github.com/pi-hole/graphics/raw/master/Vortex/Vortex_Vertical_wordmark_lightmode.png"},"container_images_update_available":true,"item_update_available":true}]}'
```

To see upgrade details like changelog mock response for `chart.release.upgrade_summary` endpoint

```
midclt call test.set_mock chart.release.upgrade_summary '{"return_value": {"container_images_to_update":{"1.0.1": "url", "1.0.2": "url2"},"changelog":"<h1>Changelog</h1>","available_versions_for_upgrade":[{"version":"15.3.36","human_version":"24.0.6_15.3.36"},{"version":"15.3.35","human_version":"24.0.6_15.3.35"},{"version":"15.3.34","human_version":"24.0.6_15.3.34"}],"item_update_available":true,"image_update_available":true,"latest_version":"15.3.36","upgrade_version":"15.3.36","latest_human_version":"24.0.6_15.3.36","upgrade_human_version":"24.0.6_15.3.36"}}'
```

To see the happy journey mock response for `chart.release.upgrade`

```
midclt call test.set_mock chart.release.upgrade '{"return_value": {"name":"true-nextcloud","version":1,"namespace":"ix-true-nextcloud","id":"true-nextcloud","catalog":"TRUECHARTS","catalog_train":"stable","path":"/mnt/tank/ix-applications/releases/true-nextcloud","dataset":"tank/ix-applications/releases/true-nextcloud","status":"ACTIVE","history":{},"update_available":true,"human_version":"24.0.6_15.3.36","human_latest_version":"24.0.6_15.3.36","pod_status":{"desired":1,"available":1},"used_ports":[],"chart_metadata":{"icon":"https://cdn.rawgit.com/docker-library/docs/defa5ffc7123177acd60ddef6e16bddf694cc35f/nextcloud/logo.svg"},"container_images_update_available":false,"item_update_available":true}}'
```

### CODE REVIEW

File: src/app/pages/applications/chart-releases/chart-releases.component.ts [hunk 1]
```diff
@@ -52,7 +53,7 @@ export class ChartReleasesComponent implements AfterViewInit, OnInit, OnDestroy
   filteredChartItems: ChartRelease[] = [];
   filterString = '';
 
-  chartItems: Record<string, ChartRelease> = {};
+  chartItems = new Map<string, ChartRelease>();
   @Output() switchTab = new EventEmitter<string>();
 
   private dialogRef: MatDialogRef<EntityJobComponent>;
```
Comment (AlexKarpov98) [inline src/app/pages/applications/chart-releases/chart-releases.component.ts:59]: can we refactor this .ts file to remove **this.dialogRef** and use const **dialogRef** instead?

**Review by undsoft — CHANGES_REQUESTED** [review]
It generally looks good. 
Some things that I've noticed:
1. When I have mouse connected, there is scrollbar in the accordion.
2. I've started an update and then opened an accordion. It showed me a weird progress bar on top.
Снимок экрана 2022-11-03 в 14 31 16

3. It seems like update status doesn't update until all apps finish updates. Would be better to show status app by app as it is coming in.
4. If possible, it would be good to show status more prominently when I start the upgrade. Like I've started the upgrade, but all I see is spinners. I have to open job manager in another tab to see % reported.
Once one of the updates failed, I had to know to hover over the icon to see the error.
Would be good to show this instead of summary in the accordion once update starts.

File: src/app/pages/applications/dialogs/chart-bulk-upgrade/chart-bulk-upgrade.component.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            this.dialogRef.close();
+          }
+        }
+      });
+      this.isJobCompleted = true;
+      this.cdr.markForCheck();
+    });
+  }
+
+  originalOrder(): number {
```
[the remark this answers is not in the document]
Comment (denysbutenko) [reply src/app/pages/applications/dialogs/chart-bulk-upgrade/chart-bulk-upgrade.component.ts:127]: No, getting errors with text: `Cannot convert undefined or null to object`

**Comment by denysbutenko** [comment]
> * It seems like update status doesn't update until all apps finish updates. Would be better to show status app by app as it is coming in.
> * If possible, it would be good to show status more prominently when I start the upgrade. Like I've started the upgrade, but all I see is spinners. I have to open job manager in another tab to see % reported.
>   Once one of the updates failed, I had to know to hover over the icon to see the error.
>   Would be good to show this instead of summary in the accordion once update starts.

3. Perhaps related to how `core.bulk` handle these requests.
4. Displaying the `core.bulk` job does not provide as much information as a single one.
image


Looking for own rxjs solution to handle bulk requests.

**Review by undsoft — APPROVED** [review]
Looks good, but if we could show minimize animation and pre-open task manager like you did somewhere else before, that would be even cooler.

**Review by AlexKarpov98 — APPROVED** [review]
That's working good, nice job!

**Pull request merged**